MM18697 - Add tests to check if tooltips show up when post menu icons are hovered

josephk96 opened the pull request:
#### Summary
I created a test that does the following : 
1. Create a new user and start a DM
2. Get post id and hover over post
3. Hover on the three post menu items/icons (dotmenu, post reaction, comment)
4. Check whether the relevant tooltip message appears when hovering over each item
5. If it does, test succeeds.

I had to make changes in the ```post_reaction.jsx``` and ```comment_icon.jsx``` files in order for Cypress to be able to hover on the respective icons. They were previously nested in a manner that did not allow Cypress to hover over the icons. Please refer to the links I have posted in the 'Related Pull Requests' section below.


#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12299
Jira: https://mattermost.atlassian.net/browse/MM-18697

thekiiingbob commented:
@josephk96 If you can, can you incorporate the required changes on the webapp in this PR? Essentially, if we submit a e2e PR, it should include whatever changes are necessary for it to work (which is usually react changes). This we we don't have to expect other PRs to be merged before the test is valid. If you do, you can close the other 2 related PRs.

josephk96 commented:
> @josephk96 If you can, can you incorporate the required changes on the webapp in this PR? Essentially, if we submit a e2e PR, it should include whatever changes are necessary for it to work (which is usually react changes). This we we don't have to expect other PRs to be merged before the test is valid. If you do, you can close the other 2 related PRs.

I incorporated the changed files into this PR and closed the other 2 PRs that I opened. Thanks for the feedback. Next time I'll make sure to incorporate all changes related to an e2e issue in 1 PR.

Below is the summary I created for the other 2 PRs that I have closed.


> I changed the order of the ```<button>``` element and the ```<OverlayTrigger>``` component within the ```comment_icon.jsx``` & the ```post_reaction.jsx``` files file so that the Cypress e2e test I wrote would be able to hover on the comment icon.

srkgupta commented:
Thanks @josephk96. Will check the latest test and let you know.

josephk96 commented:
> Hi @josephk96
> As mentioned in the JIRA ticket, can you also validate the presence of text "More Actions", "Add Reaction" and "Reply" when the tooltip appears.

Hi @srkgupta :) 
In the JIRA ticket, the expected output for the test was : 
```
*Expected:*
After 2) you should see while hovering left to right over the options, tooltips appear for:

- "More actions"

- "Add reaction"

- "Reply"
```
<img width="212" alt="Screen Shot 2019-10-10 at 11 01 03 PM" src="https://user-images.githubusercontent.com/44601888/66576552-e4645f00-ebb2-11e9-80ea-adc289581102.png">

The test that I wrote basically forced mouseovers on all three individual icons, one at a time, and checked to see whether the tooltip div appeared (the div is hidden by default when not mouseovered).

By any chance, did you mean that the icons should also be visible?

srkgupta commented:
No, I meant to validate the text of the popover shown. Something like this: 
```    
    // # Check that after hovering on the icon you would like to test, the relevant tooltip appears
    cy.get(`${tooltipName}`).should('be.visible').and('have.text',`${tooltipText}`);
```

josephk96 commented:
Ah, alright. Thanks for clearing things up

thekiiingbob commented on e2e/cypress/integration/messaging/tooltip_visual_verification_spec.js:
```diff
@@ -0,0 +1,67 @@
+    });
+
+    // # Check that after hovering on the icon you would like to test, the relevant tooltip appears
+    cy.get(`${tooltipName}`).find('span').invoke('text').then((text) => {
```
Can be shortened to  `cy.get('${toolipName}').find('span').should('have.text', tooltipText)`

josephk96 replied:
Done

thekiiingbob commented:
@josephk96 I just approved, looks like we just need a sync and merge and we should be good to go. Thanks!

thekiiingbob commented:
@josephk96 Just so it's clear, the sync and merge is something you'd need to do on your end. Let me know if you need help with it!

josephk96 commented:
@thekiiingbob I synced and merged the branch :) Sorry for not having done it earlier, I was a bit confused because I didn't have to do it for my previous PR. Thanks for helping out!
